src/gridfs/download.rs

```diff
@@ -139,16 +142,214 @@ impl GridFsBucket {
             n += 1;
         }
 
-        let expected_n =
-            total_bytes_expected / chunk_size + u64::from(total_bytes_expected % chunk_size != 0);
-        if (n as u64) != expected_n {
+        if n != file.n() {
             return Err(ErrorKind::GridFs(GridFsErrorKind::WrongNumberOfChunks {
                 actual_number: n,
-                expected_number: expected_n as u32,
+                expected_number: file.n() as u32,
             })
             .into());
         }
 
         Ok(())
     }
 }
+
+pub struct GridFsDownloadStream {
+    state: State,
+    current_n: u32,
+    cached_bytes: Option<Vec<u8>>,
+    file: FilesCollectionDocument,
+}
+
+type GetBytesFuture = BoxFuture<'static, Result<(Vec<u8>, Cursor<Chunk<'static>>)>>;
+
+enum State {
+    // Idle stores an Option<Cursor> so that the cursor can be moved into a GetBytesFuture without
+    // requiring ownership of the state. It will always store a value when the stream's state is
+    // Idle and can be unwrapped safely.
+    Idle(Box<Option<Cursor<Chunk<'static>>>>),
+    Busy(GetBytesFuture),
+    Done,
+}
+
+impl State {
+    fn set_busy(&mut self, new_future: GetBytesFuture) -> &mut GetBytesFuture {
+        *self = State::Busy(new_future);
+        match self {
+            Self::Busy(ref mut future) => future,
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl GridFsDownloadStream {
+    async fn new(
+        file: FilesCollectionDocument,
+        chunks: &Collection<Chunk<'static>>,
+    ) -> Result<Self> {
+        let initial_state = if file.length == 0 {
+            State::Done
+        } else {
+            let options = FindOptions::builder().sort(doc! { "n": 1 }).build();
+            let cursor = chunks.find(doc! { "files_id": &file.id }, options).await?;
+            State::Idle(Box::new(Some(cursor)))
+        };
+        Ok(Self {
+            state: initial_state,
+            cached_bytes: None,
+            current_n: 0,
+            file,
+        })
+    }
+
+    /// Gets the file `id` for the stream.
+    pub fn files_id(&self) -> &Bson {
+        &self.file.id
+    }
+}
+
+impl AsyncRead for GridFsDownloadStream {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<std::result::Result<usize, futures_util::io::Error>> {
+        let stream = self.get_mut();
+
+        if let Some(cache) = &mut stream.cached_bytes {
```

Comment:
Following up on my earlier comments, we can save a lot of allocations by passing this buffer to and from the future and reusing it both as the cache for leftover bytes and also as the temporary cache of bytes we're reading from chunks. This works because `Vec` is [guaranteed to never shrink its allocated capacity](https://doc.rust-lang.org/std/vec/struct.Vec.html#guarantees) even if we remove elements from it. I think I was barking up the wrong tree by suggesting limiting the cache to one chunk--we can just keep a (possibly ever-growing) cache around of arbitrary size.